core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -403,13 +533,20 @@ class LogManager(val logDirs: Array[File],
   def getLog(topicPartition: TopicPartition): Option[Log] = Option(logs.get(topicPartition))
 
   /**
-   * Create a log for the given topic and the given partition
    * If the log already exists, just return a copy of the existing log
+   * Otherwise if isNew=true or if there is no offline log directory, create a log for the given topic and the given partition
+   * Otherwise throw KafkaStorageException
+   *
+   * @param isNew Whether the replica should have existed on the broker or not
+   * @throws KafkaStorageException if isNew=false, log is not found in the cache and there is offline log directory on the broker
    */
-  def createLog(topicPartition: TopicPartition, config: LogConfig): Log = {
+  def getOrCreateLog(topicPartition: TopicPartition, config: LogConfig, isNew: Boolean = false): Log = {
     logCreationOrDeletionLock synchronized {
-      // create the log if it has not already been created in another thread
       getLog(topicPartition).getOrElse {
+        // create the log if it has not already been created in another thread
+        if (!isNew && offlineLogDirs.nonEmpty)
+          throw new KafkaStorageException(s"Can not create log for $topicPartition because log directories ${offlineLogDirs.mkString(",")} are offline")
```

Comment:
Yeah @becketqin had similar comment as well but I find that the patch in its current form is simpler. This is because not all exceptions bubble to the ReplicaManager. For example, both ReplicaFetchThread and LogClean may attempt to truncate the Log which doesn't go through ReplicaManager. And we may have more IOExeption from new methods in the future. Thus (2) alone wouldn't work.

Alternatively we can do (1) only, e.g. find all operations that may cause IOException, catch/call logDirFailureChannel.maybeAddLogFailureEvent() and re-throw a KafkaStorageException. My concern with this approach is that we will need to identify all possible calls that may throw IOException and make sure that IOException does not bubble up to ReplicaManager because otherwise we will return UnknownServerException in the response. This is doable but it requires carefully review of the code which may be error-prone both now and in the future development. On the other hand, if we can catch IOException in the ReplicaManager where we generate the error code for various response, we can ensure that if any request handling incurs IOException, we will trigger `maybeAddLogFailureEvent` and return the proper error code. Thus the current approach seems simpler and more reliable.

On the other hand I could not find concrete benefits of doing (1) only. What do you think?